python/bindings.cpp

```diff
@@ -403,6 +412,30 @@ void bindReportReader(py::module& m, const std::string& prefix) {
 
 
 PYBIND11_MODULE(_libsonata, m) {
+#if SONATA_HAS_MPI == 1
+    if (import_mpi4py() < 0) {
+        throw std::runtime_error("Could not load mpi4py API.");
+    }
+#endif
+
+    py::class_<Hdf5Reader>(m, "Hdf5Reader").def(py::init([]() { return Hdf5Reader(); }));
+
+    m.def("make_collective_reader",
+          [](py::object comm, bool collective_metadata, bool collective_transfer) {
+#if SONATA_HAS_MPI == 1
```

Comment:
When `SONATA_HAS_MPI` is False I would rather support we dont even expose collective stuff, as it might be misleading to ask for a collective reader and get a normal one.

Reply:
Summary: I think it's important that MPI is entirely optional in libsonata, even when writing code that wants to use collective MPI-IO. At most, I'd add a flag that prevents the returning a suitable default.

The lines aren't precise because likely we'll move anything MPI to its own repo. The reason is that `mpi4py` is a build dependency; and one can't have optional build dependencies (nicely).

Nevertheless, I think code that's written for collective IO should work even if the user fails to install `mpi4py`. HDF5 doesn't fail either if it can't do collective IO, instead it returns the correct values and sets an internal flag users can check. Functionally collective IO is optional. It's only important for performance reasons under certain conditions.

The risk is that installing `mpi4py` or `libsonata` with MPI support will fail to build under certain conditions. I suspect many of our users will be unable or unwilling to debug these failures. To me it's a big difference if what I'm running is a large production run that requires MPI-IO to run reasonably fast, or if it's something small (debugging) that runs fast simply because it's small. Large runs are always done on a cluster and we can expect someone adept at these issues to install the required toolchain (including MPI-enabled libsonata), even if not, there's a necessity and therefore it makes sense to spend time to debug why something failed to build. Small runs are carried out anywhere, including laptops where MPI might not be installed or it might not work correctly at that moment, etc. In my opinion I don't want to force users to have to figure these issues out just to run something that takes `2.3s` with MPI-IO and `2.2s` without MPI-IO. Similarly, I don't want every downstream project to remember to implement a graceful fallback.

To detect if the important applications are using MPI-IO we can:
* check GPFS waiters,
* run Darshan,
* use `ROMIO_PRINT_HINTS=1` and check the output, e.g., during module testing on the BB5.

Therefore, I think, on BB5, we're in a sufficiently strong position to notice if it's silently not using collective IO when it really should be.